llvm/unittests/ADT/AtomicUniquePointerTest.cpp

```diff
@@ -0,0 +1,75 @@
+//===- llvm/unittests/ADT/AtomicUniquePointerTest.cpp ---------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/ADT/AtomicUniquePointer.h"
+#include "llvm/Support/ThreadPool.h"
+#include "gtest/gtest.h"
+#include <memory>
+
+using namespace llvm;
+
+namespace {
+
+TEST(AtomicUniquePointerTest, Create) {
+  AtomicUniquePointer<int> ImplPtr;
+  SmallVector<int*, 4> Results(4);
+
+  auto createImpl = [&ImplPtr, &Results](unsigned Idx) {
+    auto Impl = std::make_unique<int>();
+    int *ExistingImpl = nullptr;
+    (void)ImplPtr.compare_exchange_strong(ExistingImpl, Impl);
+    Results[Idx] = ImplPtr.load();
+  };
+
+  ThreadPool Threads;
+  for (unsigned I = 0; I < 4; ++I)
+    Threads.async(createImpl, I);
+  Threads.wait();
+
+  EXPECT_NE(ImplPtr.load(), nullptr);
+  
+  // All the threads should get the same results.
+  SmallVector<int*, 4> ExpectedPtrs(4, ImplPtr.load());
+  EXPECT_EQ(Results, ExpectedPtrs);
+}
+
+
+TEST(AtomicUniquePointerTest, LinkList) {
+  // Test atomic insertion. Insert node S as Root -> S -> Root.Next.
+  struct DummyAtomicList {
+    AtomicUniquePointer<DummyAtomicList> Next;
+  };
+  DummyAtomicList Root;
+  auto insertNode = [&Root]() {
+    auto S = std::make_unique<DummyAtomicList>();
+    DummyAtomicList *CurrentHead = nullptr;
+    while (!Root.Next.compare_exchange_weak(CurrentHead, S))
+      S->Next.exchange(std::unique_ptr<DummyAtomicList>(CurrentHead)).release();
```

Comment:
If `compare_exchange_weak` fails _spuriously_ like it is allowed to, then `Root.Next == CurrentHead`. That would cause `X.exchange(X)` of the same pointer, which would cause UAF if the returned `unique_ptr` is destroyed without releasing.

Reply:
... also no, if it failed non-spuriously, the returned unique-ptr would be pointing to something in the main list, so again you don't want to delete it.

Reply:
Seems to me that a `unique_ptr<>`-based API that requires a bunch of `.release()` calls needs some iteration on the design of the API.

But `compare_exchange_weak()` has odd implications for ownership. Would be good to see some of the other APIs tested in isolation; maybe it's mostly okay.

Reply:
Yeah, it looks all those `.release()` is needed for not freeing up the item in the list.

Reply:
Looking at few options. I think the weak version of compare_exchange is really bad for unique_ptr style objects.